src/approximate-qvm.lisp

```diff
@@ -0,0 +1,173 @@
+        :do (cond ((= (list-length qubits) 1) 
+                   (let ((kraus-map (gethash (nth 0 qubits) source)))
+                     (when kraus-map (progn (check-kraus-ops kraus-map)
+                                            (apply-kraus-ops qvm instr kraus-map)))))
+                  ((= (list-length qubits) 2) 
+                   (let ((kraus-map  (kraus-kron (gethash (nth 0 qubits) source) 
+                                                 (gethash (nth 1 qubits) source))))
+                     (when kraus-map (progn (check-kraus-ops kraus-map)
+                                            (apply-kraus-ops qvm instr kraus-map)))))              
+                  ((> (list-length qubits) 2) (format t "error -- cannot have > 2 qubits"))))) ; TODO actual error check
+
+
+(defmethod transition ((qvm approx-qvm) (instr quil:gate-application))
+  "For the current instruction in the program, apply the gate corresponding to that
+instruction, and then apply all the kraus operators for all the sources of noise following
+that instruction. "
+  (let ((gate   (pull-teeth-to-get-a-gate instr))
+        (params (mapcar (lambda (p) (force-parameter p qvm))
+                        (quil:application-parameters instr)))
+        (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr))))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple qubits) params)
+    (apply-all-kraus-operators qvm instr qubits)
+    (incf (pc qvm))
+    qvm))
+
+
+(defun tphi (t1 t2)
+  "Calculate t_phi from t1 and t2.
+   t_phi = (2*t1*t2) / (2*t1 + t2)"
+  (/ (* 2 (* t1 t2)) (+ t2 (* 2 t1))))
+
+
+(defun generate-damping-kraus-ops (t1 gate-time)
+  "Given a value for t1 and a gate time, generates the kraus operators corresponding
+to the t1 noise. "
+  (let* ((prob (/ gate-time t1)) 
+         (k0 (magicl:make-complex-matrix 2 2 (list 0 0 (sqrt prob) 0)))
+         (k1 (magicl:make-complex-matrix 2 2 (list 1 0 0  (sqrt (- 1 prob))))))
+    (list k0 k1)))
+
+
+(defun generate-dephasing-kraus-ops (tphi gate-time)
+  "Given a value for t_phi (dephasing time) and a gate time, calculates the kraus
+operators corresponding to the dephasing noise."
+  (let*  ((prob (/ gate-time tphi))
+          (prob-k1 (/ prob 2))
+          (prob-k0 (- 1 prob-k1 ))
+          (kraus-ops (loop :for mat :in '("I" "Z")
+                           :for pj :in (list prob-k0 prob-k1)
+                           :collect (magicl:scale (sqrt pj)
+                                                  (quil:gate-matrix
+                                                   (quil:gate-definition-to-gate
+                                                    (quil:lookup-standard-gate mat)))))))
+    kraus-ops
+    )) 
+
+(defun generate-damping-dephasing-kraus-ops (t2 gate-time)  
+  "Calculates the kraus operators for dephasing and damping noise from t2, which
+is decoherence time. "
+  (generate-dephasing-kraus-ops t2 gate-time)) 
```

Comment:
Yeah -- I am pretty sure the only difference between `damping-dephasing` and `dephasing` is that pure `dephasing` uses `tphi` and `damping-dephasing` uses t2. I made two separate methods because I wanted to make absolutely sure that the calculations are the same before combining them into the same function. I'll add a clarifying comment for `damping-dephasing` while I do some more research to figure that out